### PR TITLE
MSC2730: Verifiable forwarded events

### PR DESCRIPTION
**Pull request by tulir**
[Rendered](https://github.com/tulir/matrix-doc/blob/verifiable-forwarded-events/proposals/2730-verifiable-forwarded-events.md)

This is an alternative to #2723

### CODE REVIEW

File: proposals/2730-verifiable-forwarded-events.md [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+a notice saying it's unverified.
+
+## Potential issues
+* This is not as simple as MSC2723 and requires server support.
+* Events with bogus data in `m.forwarded` can't be forwarded.
+* Events that are close to the 64 KiB size limit can't be forwarded. MSC2723
+  has the same problem, but this proposal has even more extra data. The amount
+  of extra data in both proposals is rather low (<1kb), so this should not be
+  a problem in practice.
+* Encrypted events forwarded to other rooms won't be decryptable. Clients
```
Comment (Sorunome) [inline proposals/2730-verifiable-forwarded-events.md:234]: unfortunatley this is a hard-blocker for our usecase. We need that messages can be forwarded both from and to encrypted rooms.

**Comment by tulir** [comment]
There's now an initial Synapse implementation: matrix-org/synapse#8078

File: proposals/2730-verifiable-forwarded-events.md (hunk 1 above)
Comment (tulir) [reply proposals/2730-verifiable-forwarded-events.md:235]: I don't think there's any way to have verifiable forwards of encrypted messages, except with other changes like message signing by clients or sharing the original message keys.

Using both proposals could maybe work: verifiable forwards for unencrypted events, user-set metadata for encrypted events

Comment (reivilibre) [reply proposals/2730-verifiable-forwarded-events.md:235]: Yes, this would be my main concern as well. I frequently forwarded message in and out of encrypted rooms when I was using a client with that feature.

I don't know enough about megolm, but if we have per-message keys, would it be possible to disclose those?

Hum :/

Comment (tulir) [reply proposals/2730-verifiable-forwarded-events.md:235]: Currently keys are not per-message, but if they were, the client could just share those with everyone in the target room.

**Comment by MurzNN** [comment]
Maybe allow to forward multiple events in one forwarded event? This feature will can also partly implement https://github.com/matrix-org/matrix-doc/issues/2289

File: proposals/2730-verifiable-forwarded-events.md (hunk 1 above)
Comment (tulir) [reply proposals/2730-verifiable-forwarded-events.md:235]: So I just noticed that the AES keys used to encrypt messages are actually technically per-message: they're derived from the megolm session using HKDF, so I think you can share the resulting key without compromising the whole session.

It would definitely require additional client-side implementation to be able to decrypt the message using such keys, but that's better than not being able to forward encrypted messages at all.

File: proposals/2730-verifiable-forwarded-events.md [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+However, only keys for the message being forwarded should be sent, any other
+messages in the origin room must not be decryptable with those keys.
+
+To achieve this, the user forwarding the message includes the message-specific
+symmetric AES and HMAC keys (see [Message encryption] in the Megolm spec). Each
+of the keys are encoded as unpadded base64 and placed in the `aes_key`,
+`hmac_key` and `aes_iv` fields in the `decryption_keys` object in the forward
+request.
+
+[Message encryption]: https://gitlab.matrix.org/matrix-org/olm/-/blob/master/docs/megolm.md#message-encryption
```
Comment (tulir) [inline proposals/2730-verifiable-forwarded-events.md:234]: Someone who knows cryptography should make sure this actually makes sense and is secure enough (i.e. doesn't leak anything else than the one message being forwarded)

File: proposals/2730-verifiable-forwarded-events.md [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+      },
+      "signatures": {
+        "maunium.net": {
+          "ed25519:a_xxeS": "cc9XnH9ByO7yadC6CdMhh3c/TN1tQ9FiZdKYyRDi4Og1dZMylmBM9uSI7c4GUEqswLBLxW5DTFU3n7vMHAGhAw"
+        }
+      },
+      "unsigned": {
+        "displayname": "tulir",
+        "avatar_url": "mxc://maunium.net/jdlSfvudiMSmcRrleeiYjjFO"
+      }
```
Comment (auscompgeek) [inline proposals/2730-verifiable-forwarded-events.md:172]: This example is missing the `room_version`. (Relatedly, what should happen if it is missing?)

File: proposals/2730-verifiable-forwarded-events.md [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+explicit `"valid": true` in the unsigned `m.forwarded` object. Additionally,
+clients SHOULD make sure the server supports this proposal before trusting
+forwards even if the `valid` flag is present.
+
+Not trusting forward metadata does not necessarily mean it must be completely
+ignored. For example, clients could render the event as a forward, but include
+a notice saying it's unverified.
+
+When receiving forwarded encrypted events, clients should treat the message
+like they treat forwarded keys, i.e. not confirmed to originate from the user.
```
Comment (kevincox) [inline proposals/2730-verifiable-forwarded-events.md:256]: To be clear what should the assume? AKA what does this verification give us?

IIUC it tells us if our homeserver trusts that this event came from the homeserver of the sender? Is this correct?

File: proposals/2730-verifiable-forwarded-events.md [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+Clients may discourage users from forwarding encrypted messages to unencrypted
+rooms, as that would leak the message content to the servers.
+
+If a forwarded event contains relation metadata such as a reply, clients should
+not display it to users. This behavior is consistent with other platforms (e.g.
+Telegram and WhatsApp) and removes any problems if some users can't get the
+relation target event. The existence of reply metadata may still be used to
+remove reply fallbacks.
+
+## Potential issues
```
Comment (kevincox) [inline proposals/2730-verifiable-forwarded-events.md:267]: A concern I have is that this requires forwarding all of the information in the original event. I can imagine cases where I may want to reference the author, but not expose the room. Other cases could not want to reveal the time of the message. Basically there is a legitimate use case for including any subset of the information from the original event.

Furthermore it may be unclear to the forwarded exactly what information they are sharing.

This proposal seems to suggest that it is not allowed to forward modified (including redacted) events. Of course modified events can not be verified but I think it is still nice to have all of this structured data about the event even if it can't be verified to be coming from the original user's homeserver. For example I may just want to forward (message, author, time) and am perfectly fine saying "according to Kevin".

So basically how should we handle non-verifiable forwarded events? The same as this protocol but just without `"valid": true`? Or do we want to separate spec?

File: proposals/2730-verifiable-forwarded-events.md (hunk 3 above)
Comment (kevincox) [reply proposals/2730-verifiable-forwarded-events.md:172]: To expand on this what if the room version is not supported by clients? I guess in most cases it will *just work* but it seems unfortunate if a client can't see a forward in a room with a supported version because the event is from a newer versioned room.

Comment (tulir) [reply proposals/2730-verifiable-forwarded-events.md:172]: Clients don't need to support the room version, but servers do. I guess missing or unsupported `room_version` can just be marked as non-trusted/invalid

File: proposals/2730-verifiable-forwarded-events.md [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+## Proposal
+The proposed solution is copying the entire federation event data, which allows
+recipients to validate the signatures even if they are not in the origin room.
+
+As clients generally don't have access to signatures nor any way to validate
+them, both sending and validating require server support. Sending is
+implemented as a new endpoint, while validating happens automatically and the
+server adds the validation result to the top-level `unsigned` object.
+
+### `PUT /_matrix/client/r0/rooms/{roomId}/event/{eventId}/forward/{targetRoomId}/{txnId}`
```
[the remark this answers is not in the document]
Comment (tulir) [reply proposals/2730-verifiable-forwarded-events.md:14]: The `PUT` is there mostly because of `{txnId}`, otherwise it would be `POST`.

The current path is that way because you're doing something with the event. `/_matrix/client/r0/rooms/{targetRoomId}/forward/{sourceRoomId}/{eventId}/{txnId}` feels less clear in that way

[the remark this answers is not in the document]
Comment (kevincox) [reply proposals/2730-verifiable-forwarded-events.md:14]: But you aren't actually doing something to the event. You are creating a new event that is based on it and references it.

File: proposals/2730-verifiable-forwarded-events.md (hunk 3 above)
Comment (kevincox) [reply proposals/2730-verifiable-forwarded-events.md:172]: That isn't a great experience. If I am in a room version X why am I being sent content in a different room version? Maybe we should support some sort of fallback content? Of course now you need to worry about verifying that the fallback content matches the source content. (which is basically impossible if the room version the source is from us unsupported)

Comment (tulir) [reply proposals/2730-verifiable-forwarded-events.md:172]: The only client-facing change in room versions so far is the event ID (which is mentioned in the proposal). Even in the future there most likely won't be any huge client-facing changes, since room versions are for changing the federation protocol. Major client-facing changes like extensible events won't bump the room version.

File: proposals/2730-verifiable-forwarded-events.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
+# Verifiable forwarded events
+This is an alternative to [MSC2723](https://github.com/matrix-org/matrix-doc/pull/2723)
+that handles the issue of faking forwards.
```
Comment (kevincox) [inline proposals/2730-verifiable-forwarded-events.md:3]: Thinking about this more I'm not sure that this represents a valuable use case. I also think the risk of forwarding more information than you wanted to is too high.

Particularly for the best forwarding experience it probably makes sense to trim the message anyways.  If we want to support various ways of redacting parts of the forwarded event (which I think we should) we currently don't have a way to verify that and it will be confusing to clients to warn about "according to the sender" for some forwards but having other forwards be verified.

The most compelling use case I can think of are abuse reports but I think in most cases the moderator receiving the report can see the room anyways or otherwise check the legitimacy of the report.

Comment (MurzNN) [reply proposals/2730-verifiable-forwarded-events.md:3]: Forwarding message means that it isn't changed, if you modify the message text (redact part of it), it is already quoting, not forwarding.

**Review by ShadowJonathan — APPROVED** [review]
All in all i find this a sound proposal with some genius ideas, i'd still recommend someone versed in cryptography and privacy to look at all the data this is forwarding as well, and if it's possible to redact any of it (optionally)

The prospect of adding another endpoint to a server would make me not like this MSC on first glance (in a world where P2P is the future), but seeing as it's merely needed to have the server copy signature information (which is then self-verifying by other servers/devices), I like this solution.

File: proposals/2730-verifiable-forwarded-events.md (hunk 5 above)
Comment (jhulkko) [reply proposals/2730-verifiable-forwarded-events.md:267]: In my opinion copying content from a message to other selectively is a totally different use-case than forwarding a full message. This MSC should be tightly restricted to forwarding full events only, and more importantly, **I would explicitly deny forwarding of encrypted events to un-encrypted rooms!**

When forwarding an encrypted event, you don't only copy the original message, but the actual point-in-time event sent by an original author and including all the meta information in it to another context. While this probably has a bunch of valid use-cases, I consider it a very bad practice to make it easy for 2nd parties of original communication to leak all this to the HSes when the original message content is deliberately shielded from server side spying. When a user copy-pastes a content of a originally encrypted message, the relation to the original encrypted counterpart is not explicitly leaked to the homeservers, and the relation to the original encrypted context is not explicitly crytpographically proven. This is a major difference that should be considered when forwarding the original events to other contexts.

Thus, my recommendation is that forwarding of encrypted messages to unencrypted rooms should be blocked by both the server and the client implementations explicitly. Forwarding of encrypted events to other encrypted rooms does not break the separation of trust contexts between the HS and the clients as long as the decryption keys are encrypted with the keys of the target room.

Regarding the thoughts on selectively "forwarding" some of the event information the current signing scheme does not support being picky on what to include or exclude if you want to be able to verify the signature. Building something that would allow forwarding selectively, while still providing cryptographic proofs would require a hugely more complex cryptosystem design than Matrix currently has. One possibility to that direction would be using something like Zero Knowledge proofs, but that is totally out of scope in terms of added complexity for this single use case.

File: proposals/2730-verifiable-forwarded-events.md (hunk 2 above)
Comment (jhulkko) [reply proposals/2730-verifiable-forwarded-events.md:234]: When forwarding events to other encrypted rooms this approach does not seem to leak any additional information nor session security related secrets of the megolm context of the originating room. The receiving room members can only encrypt this single message under normal circumstances. If there are any corner cases or bugs in the clients that cause re-use of the megolm index for two events, then that specific event would be decryptable as well, but this should not happen anyway.

The actual issue to consider is that should we break the trust model of Matrix E2EE by leaking the key information to the HS? What makes this a big issue is that the actual original event contents combined with all of the event meta information is then explicitly leaked outside the original E2EE context in cryptographically verifiable manner. This is totally different case than copy-pasting secret message to another context and claiming that it originated somewhere.

Thus, I would strongly suggest that E2EE secured events should only be allowed to be forwarded in this verifiable manner to other E2EE rooms.

Comment (tulir) [reply proposals/2730-verifiable-forwarded-events.md:234]: I don't think there should be mandatory server-side limits, but I guess I could reword the client section to say "Clients are strongly recommended to discourage or disallow users from forwarding encrypted messages to unencrypted rooms".

(also, megolm keys are technically not cryptographically verifiable, only keys received via a secure olm channel can actually be verified to originate from the user - this is mentioned in the client behavior section)

Comment (jhulkko) [reply proposals/2730-verifiable-forwarded-events.md:234]: This does still build a direct functionality to the spec to leak the message related secrets to the HS in plaintext, if forwarding to unencrypted rooms is allowed. HS itself has the original event, and thus can then see not only the meta information of the event it originally was aware of, but also the contents of it.

IMO this approach would be a violation of the implicit trust model Matrix tries to provide by E2EE. This design decision should not be made lightly.

Would it be worth considering to leave the decision of event forwarding to the context of a room? Then forwarding could be allowed to other encrypted rooms only, or all rooms, or denied in full. This will at least provide some guarantees that the data hygiene and separation of privileges can be upheld to sufficient degree.

File: proposals/2730-verifiable-forwarded-events.md [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+* `signatures`
+
+The following keys may also be present:
+
+* `prev_state`, may be present as an empty array even in non-state events
+* `event_id`, only in v1 and v2 rooms
+
+Additionally, the server MUST include an `unsigned` object, containing a
+`room_version` field that specifies the version of the source room. The server
+SHOULD also include the sender's profile metadata in the unsigned object under
```
Comment (ShadowJonathan) [inline proposals/2730-verifiable-forwarded-events.md:73]: Isn’t this unsigned/unverified profile data at risk of falsification?
